### PR TITLE
supplemental data between compacting and commit compacting

### PR DESCRIPTION
**Pull request by hxiaodon**
Chris:

This is the draft code for issue [374](https://github.com/chrislusf/seaweedfs/issues/374), please help to review and give me feedback.
Thanks!

### CODE REVIEW

File: weed/storage/volume.go [hunk 1]
```diff
@@ -23,6 +28,9 @@ type Volume struct {
 
 	dataFileAccessLock sync.Mutex
 	lastModifiedTime   uint64 //unix time in seconds
+
+	lastCompactingIndexOffset       uint64
+	incrementedHasUpdatedIndexEntry map[uint64]keyField
```
Comment (chrislusf) [inline weed/storage/volume.go]: incrementedHasUpdatedIndexEntry does not need to be in this struct.

Comment (hxiaodon) [reply weed/storage/volume.go]: Yeah, I know current garbage collection is ordered(one by one volume). But I have the concern that when concurrent compacting is involved in weed some day, these 2 fields should become volume specific.

**Comment by hxiaodon** [comment]
the old index offset directly writing into the new index  has something, the new dat file's offset should be used. Anyway, I will modify the code finally after consolidate all you comments

File: weed/storage/volume.go (hunk 1 above)
Comment (chrislusf) [reply weed/storage/volume.go]: I am only talking about incrementedHasUpdatedIndexEntry. The map only used locally in makeupDiff()

File: weed/storage/volume_vacuum.go [hunk 2]
```diff
@@ -55,7 +71,85 @@ func (v *Volume) commitCompact() error {
 	return nil
 }
 
-func makeupDiff(newDatFile, newIdxFile, oldDatFile, oldIdxFile string) (err error) {
+func (v *Volume) makeupDiff(newDatFileName, newIdxFileName, oldDatFileName, oldIdxFileName string) (err error) {
```
Comment (chrislusf) [inline weed/storage/volume_vacuum.go:90]: How about adding a test case for this function?

Comment (hxiaodon) [reply weed/storage/volume_vacuum.go:90]: Is there some good ways to do it? I 'm thinking about creating 4 test volume/index files in advance(new creating, updating and deleting appended data should be in them), then write a test case  with these volume files as input

File: weed/storage/volume.go (hunk 1 above)
Comment (hxiaodon) [reply weed/storage/volume.go]: Got it, my local code has been modified

File: weed/storage/volume_vacuum.go (hunk 2 above)
Comment (chrislusf) [reply weed/storage/volume_vacuum.go:90]: yes. this would work.

**Comment by hxiaodon** [comment]
Chris,
Please take a review of my UT case. Since the UT input files differ one from each other, so I add the comments for it and comment the UT case
And I did run the ut case at my local, fix some bugs(such as fakedDelNeedle.) Please help to check my fix and give your feedback
Thanks!

**Pull request merged**